src/field/group.rs

```diff
@@ -0,0 +1,135 @@
+//! Groups $G$ are algebraic structures which are set and has a binary operation $\oplus$ that
+//! combines two elements $a, b$ of the set to produce a third element $a\oplus b$ in the set.
+//! The operation is said to have following properties:
+//! 1. Closure: $a\oplus b=c\in G$
+//! 2. Associative: $(a\oplus b)\oplus c = a\oplus(b\oplus c)$
+//! 3. Existence of Identity element: $a\oplus 0 = 0\oplus a = a$
+//! 4. Existence of inverse element for every element of the set: $a\oplus b=0$
+//! 5. Commutativity: Groups which satisfy an additional property: *commutativity* on the set of
+//!    elements are known as **Abelian groups**.
+//!
+//! [`FiniteGroup`] trait is implemented for all finite groups.
+use std::ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use super::{prime::find_primitive_element, FiniteField, PrimeField};
+
+#[const_trait]
+/// Group trait with finite number of elements
+pub trait FiniteGroup:
+  std::fmt::Debug
+  + Sized
+  + Add<Output = Self>
+  + AddAssign
+  + Sub<Output = Self>
+  + SubAssign
+  + Neg
+  + Mul<Self::Scalar, Output = Self>
+  + MulAssign<Self::Scalar>
+  + Clone
+  + Copy
+  + Default
+  + Eq {
+  /// Scalar defined in the group
+  type Scalar: FiniteField;
```

Comment:
This is confusing to me.

Types used in a group could realistically be anything. I think this is okay if it is a curve group, but would prefer to not typecast a general `FiniteGroup` to operate on `FiniteField`

For instance, consider the abstract (dihedral) group (presentation) `<r,f | r^n, f^2, rfrf>`. One could represent this group with two chars (or an enum with two variants) and define the operations via reduction on these rules.

Reply:
You're right. Changed it now to abstract type which can be defined by the implementation.